src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ViewBuffer.cs

```diff
@@ -254,5 +210,139 @@ public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+                    {
+                        destination.AppendHtml((IHtmlContent)value.Value);
+                    }
+                }
+            }
+        }
+
+        public void MoveTo(IHtmlContentBuilder destination)
+        {
+            if (Pages == null)
+            {
+                return;
+            }
+
+            var other = destination as ViewBuffer;
+            if (other != null)
+            {
+                MoveTo(other);
+                return;
+            }
+
+            for (var i = 0; i < Pages.Count; i++)
+            {
+                var page = Pages[i];
+                for (var j = 0; j < page.Count; j++)
+                {
+                    var value = page.Buffer[j];
+
+                    string valueAsString;
+                    IHtmlContentContainer valueAsContainer;
+                    if ((valueAsString = value.Value as string) != null)
+                    {
+                        destination.AppendHtml(valueAsString);
+                    }
+                    else if ((valueAsContainer = value.Value as IHtmlContentContainer) != null)
+                    {
+                        valueAsContainer.MoveTo(destination);
+                    }
+                    else
+                    {
+                        destination.AppendHtml((IHtmlContent)value.Value);
+                    }
+                }
+            }
+
+            for (var i = 0; i < Pages.Count; i++)
+            {
+                var page = Pages[i];
+                Array.Clear(page.Buffer, 0, page.Capacity);
+                _bufferScope.ReturnSegment(page.Buffer);
+            }
+
+            Pages.Clear();
+        }
+
+        private void MoveTo(ViewBuffer destination)
+        {
+            for (var i = 0; i < Pages.Count; i++)
+            {
+                var page = Pages[i];
+
+                var destinationPage = destination.Pages.Count == 0 ? null : destination.Pages[destination.Pages.Count - 1];
+
+                // If the other page is less or equal to than half full, let's copy it's to the current page if
+                // possible.
+                var isLessThanHalfFull = 2 * page.Count <= page.Capacity;
+                if (isLessThanHalfFull &&
+                    destinationPage != null &&
+                    destinationPage.Capacity - destinationPage.Count >= page.Count)
+                {
+                    // We have room, let's copy the items.
+                    Array.Copy(
+                        sourceArray: page.Buffer,
+                        sourceIndex: 0,
+                        destinationArray: destinationPage.Buffer,
+                        destinationIndex: destinationPage.Count,
+                        length: page.Count);
+
+                    destinationPage.Count += page.Count;
+
+                    // Now we can return this page, and it can be reused in the scope of this request.
+                    _bufferScope.ReturnSegment(page.Buffer);
```

Comment:
Need `Array.Clear()` here for same reason as you've got one in the other `MoveTo()` method.